proposed/2022/PrivateAssetIndepdentFromExcludeAsset.md

```diff
@@ -0,0 +1,147 @@
+# Opt-in `PrivateAssets` asset flow control option precedence over `IncludeAssets/ExcludeAssets` option
+
+- Author Name <https://github.com/erdembayar>
+- Start Date 2022-12-07
+- GitHub Issue <https://github.com/NuGet/Home/issues/6938>
+- GitHub PR <https://github.com/NuGet/NuGet.Client/pull/4976>
+
+## Summary
+
+<!-- One-paragraph description of the proposal. -->
+Currently `IncludeAssets/ExcludeAssets` options take precedence over `PrivateAssets` asset control option when creating package, so in many cases `IncludeAssets/ExcludeAssets` options completely eclipse the `PrivateAssets` option so doesn't let assets flow to consuming parent project. This proposal introduces new a `PackPrivateAssetsFlow` opt-in property so it let `PrivateAssets` asset control option takes precedence over `IncludeAssets/ExcludeAssets` options to enable asset flow to consuming parent projects.
```

Comment:
Personally, I find this summary confusing, and doesn't convey what I currently understand the feature actually being about.  Maybe I still don't understand the feature proposal correctly, but here's what it currently is in my mind:

```suggestion
The `IncludeAssets`, `PrivateAssets` and `ExcludeAssets` metadata on `PackageReference` items control two different features. Firstly, which assets from a package that are included in the current project. Secondly, whether the assets will be listed in the package's dependency assets, for those assets to flow transitively, if the project is packed.

For example, `<PackageReference Include="Microsoft.SourceLink.GitHub" Version="1.0.0" PrivateAssets="all" />` means "Include the default assets in the current project, but if packed, all of the assets are excluded, so this package will not be a dependency".

Another example, `<PackageReference Include="NuGet.Protocol" Version="6.4.0" PrivateAssets="compile" />` means "include the default assets in the current project, but if packed, the "compile" asset should be excluded from the package dependency, so that my package's dependencies do not leak APIs into projects using my package".

A third example, `<PackageReference Include="Microsoft.Build" Version="17.0" ExcludeAssets="runtime" />` means "in the current project, make the `compile` assets available (so both intellisense and the compiler can access APIs from the package), but `Microsoft.Build`'s dlls are not copied to `bin` on build (`runtime` assets), and if the project is packed, then `runtime` assets also be excluded for the `Microsoft.Build` dependency for any project that references the current project's package.

However, a scenario that is missing is "exclude a package asset from a current project, but do not exclude it from the dependency when the current project is packed". This missing feature is more obvious when looking at a project/flow matrix:

||Flow transitively|Don't flow transitively|
|--|--|--|
|**Include in project**|IncludeAssets|PrivateAssets|
|**Don't include in project**||ExcludeAssets|
```

Reply:
I like your description and examples. The table you end with is less clear. You can include/exclude using either IncludeAssets or ExcludeAssets (just by listing the complement instead). PrivateAssets does *not* control what gets included in the project.
I think it _would_ be useful though to have such a truth table that shows everything about how these metadata can overlap (in particular, how Include/Exclude can influence/override what you set in PrivateAssets) in the default mode, and how the property changes this.

Reply:
@zivkan 
Thank you for great write up. I applied your suggestion.

@AArnott 
Could you check the modified table I added here?

Reply:
Just plugging https://github.com/NuGet/Home/pull/12313#discussion_r1044703269 :)

Reply:
> Just plugging [#12313 (comment)](https://github.com/NuGet/Home/pull/12313#discussion_r1044703269) :)

Yes, great summary package consumption case. In addition, to that we're considering for P2P scenario too.

```
I think the best description of IncludeAssets/ExcludeAssets and PrivateAssets is in https://github.com/NuGet/docs.microsoft.com-nuget/issues/2716#issuecomment-1092223183. 

As a project, `IncludeAssets` and `ExcludeAssets` indicate what I consume from a package. (EffectiveInclude)
As a project, `PrivateAssets` (today in combination with EffectiveInclude) indicates what projects that reference me consume. 

The ask is that EffectiveInclude is not combined with PrivateAssets
```